### PR TITLE
Fix issue#37 Add TR to PL words on QuestionWordResponse

### PR DESCRIPTION
**Pull request by hamdiboyraz**

### CODE REVIEW

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 1]
```diff
@@ -129,17 +129,31 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
                                    {
                                        Word = x.Word,
                                        Id = x.Id,
                                        WritingInLanguage = x.WritingInLanguage,
                                        Meaning = x.Meaning
                                    }).ToListAsync();
+            
+            var randomWordsTRtoPL = await Queryable().Where(x => !x.IsLearned)
+                .OrderBy(x => Guid.NewGuid())
+                .Take(20-randomNumber)
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:147]: take "20-randomNumber" to a variable

**Review by Jerolivine — COMMENTED** [review]
image

please check

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 2]
```diff
@@ -129,17 +129,31 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
 
         public async Task<List<QuestionWordResponse>> GetQuestionWordsAsync()
         {
-            var randomWords = await Queryable().Where(x => !x.IsLearned)
+            int randomNumber = new Random().Next(0, 21);
+
+            var randomWordsPLtoTR = await Queryable().Where(x => !x.IsLearned)
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:134]: await Queryable().Where(x => !x.IsLearned).OrderBy(x => Guid.NewGuid())

this line is used two times. take it to a single variable and use it for both queries.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 3]
```diff
@@ -129,17 +129,31 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
                                    .OrderBy(x => Guid.NewGuid())
-                                   .Take(20)
+                                   .Take(randomNumber)
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:136]: randomNumber is not meaningful.

File: Back End/MemorizeWords/MemorizeWords/Presentation/Models/Response/QuestionWordResponse.cs [hunk 4]
```diff
@@ -6,5 +6,6 @@ public class QuestionWordResponse
         public string Word { get; set; }
         public string WritingInLanguage { get; set; }
         public string Meaning { get; set; }
+        public int Language { get; set; }
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Presentation/Models/Response/QuestionWordResponse.cs:9]: LanguageType

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 5]
```diff
@@ -129,17 +129,31 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
                                    .Select(x => new QuestionWordResponse()
                                    {
                                        Word = x.Word,
                                        Id = x.Id,
                                        WritingInLanguage = x.WritingInLanguage,
                                        Meaning = x.Meaning
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:142]: Set LanguageType

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 6]
```diff
@@ -129,17 +129,31 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
                                    }).ToListAsync();
+            
+            var randomWordsTRtoPL = await Queryable().Where(x => !x.IsLearned)
+                .OrderBy(x => Guid.NewGuid())
+                .Take(20-randomNumber)
+                .Select(x => new QuestionWordResponse()
+                {
+                    Word = x.Meaning,
+                    Id = x.Id,
+                    Meaning = x.Word
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:152]: Set LanguageType

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Constants/AppSettings/LanguageType.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace MemorizeWords.Infrastructure.Constants.AppSettings;
+
+public enum LanguageType
+{
+    LearningLanguage = 0,
+    MyLanguage = 1
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Constants/AppSettings/LanguageType.cs:6]: UserLanguage=1

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs (hunk 1 above)
Comment (Jerolivine) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:147]: randomWordsTRtoPL 

TR to PL olmasın adı. İleride kullanicilar kendi dilinden baska dile kullanacak. Genel bir isim vermeyi dene

Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs (hunk 2 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs (hunk 3 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Presentation/Models/Response/QuestionWordResponse.cs (hunk 4 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Presentation/Models/Response/QuestionWordResponse.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs (hunk 5 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs (hunk 6 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Constants/AppSettings/LanguageType.cs (hunk 7 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Constants/AppSettings/LanguageType.cs]: Yapıldı.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 8]
```diff
@@ -129,16 +129,33 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
-                                   .Take(20)
-                                   .Select(x => new QuestionWordResponse()
-                                   {
-                                       Word = x.Word,
-                                       Id = x.Id,
-                                       WritingInLanguage = x.WritingInLanguage,
-                                       Meaning = x.Meaning
-                                   }).ToListAsync();
+            var learningLanguageWordCount = new Random().Next(0, 21);
+            var userLanguageWordCount = 20 - learningLanguageWordCount;
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:133]: 20 is magic string. make it private const int value under WordRepository class.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 9]
```diff
@@ -129,16 +129,33 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
-                                   .OrderBy(x => Guid.NewGuid())
-                                   .Take(20)
-                                   .Select(x => new QuestionWordResponse()
-                                   {
-                                       Word = x.Word,
-                                       Id = x.Id,
-                                       WritingInLanguage = x.WritingInLanguage,
-                                       Meaning = x.Meaning
-                                   }).ToListAsync();
+            var learningLanguageWordCount = new Random().Next(0, 21);
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:132]: new Random().Next(0, 21)

make it method and give a meaningful name and use 20 value from constant value

**Review by Jerolivine — CHANGES_REQUESTED** [review]

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs [hunk 10]
```diff
@@ -129,16 +129,33 @@ private static void ValidationAddUpdateWord(WordAddRequest wordAddRequest)
...
+            var userLanguageWordCount = 20 - learningLanguageWordCount;
+
+            var baseQuery = Queryable().Where(x => !x.IsLearned).OrderBy(x => Guid.NewGuid());
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordRepository.cs:135]: baseQuery is not meaningfulname. It doesnt tell any meaning.

**Review by Jerolivine — CHANGES_REQUESTED** [review]

File: Back End/MemorizeWords-UnitTest/UnitTest.cs [hunk 11]
```diff
@@ -82,8 +84,28 @@ public void Answer_ShouldAddRecord()
...
         {
             return new DbContextOptionsBuilder<EFCoreDbContext>()
-                            .UseInMemoryDatabase(databaseName: "TestDatabase")
-                            .Options;
+                .UseInMemoryDatabase(databaseName: "TestDatabase")
+                .Options;
+        }
+
+        [Fact]
+        public void IS_DEFAULT_TEST_SHOULD_DONE()
```
Comment (Jerolivine) [inline Back End/MemorizeWords-UnitTest/UnitTest.cs:92]: this method looks great.

Can you move this method to new UnitTest class?

Create a new Folder called Utility and Create a class called ValidationUtilityTests.cs and put this method under it 

Great!

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs [hunk 12]
```diff
@@ -9,16 +10,19 @@ public static T GetSettingsValue<T>(this IConfiguration configuration, string se
         {
             var value = configuration.GetValue<T>(settingsName);
 
-            if (EqualityComparer<T>.Default.Equals(value, default(T)))
-            {
-                throw new BusinessException($"{settingsName} is not an integer");
-            }
+            
+            // if (EqualityComparer<T>.Default.Equals(value, default(T)))
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs:14]: Delete comments

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs [hunk 13]
```diff
@@ -9,16 +10,19 @@ public static T GetSettingsValue<T>(this IConfiguration configuration, string se
...
+            // {
+            //     throw new BusinessException($"{settingsName} is not an integer");
+            // }
+            
+            ValidationTypeUtility.ThrowIfNullOrDefault(value, settingsName);
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs:19]: Give exception message as parameter

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordAnswerRepository.cs [hunk 14]
```diff
@@ -94,7 +95,18 @@ private bool GetGivenAnswer(WordAnswerRequest wordAnswerRequest, out WordEntity
         {
             wordEntity = _dbContext.Word.FirstOrDefault(x => x.Id == wordAnswerRequest.WordId) ?? throw new KeyNotFoundBusinessException($"wordId: {wordAnswerRequest.WordId} couldn't found");
 
-            bool answer = wordEntity.Meaning.ToUpperInvariant().Equals(wordAnswerRequest.GivenAnswerMeaning.ToUpperInvariant().ToUpper(), StringComparison.OrdinalIgnoreCase);
+            string targetLanguage = wordEntity.Meaning;
+            
+            if ((LanguageType)wordAnswerRequest.AnswerLanguageType == LanguageType.LearningLanguage)
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordAnswerRepository.cs:100]: Make this if else blocks a method. But the name can be GetCrossCheckWord(wordEntity)

Comment (Jerolivine) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordAnswerRepository.cs:100]: If you see a code block for only a purpose, always extract method.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Utilities/ValidationTypeUtility.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿using MemorizeWords.Infrastructure.Transversal.Exception.Exceptions;
+
+namespace MemorizeWords.Infrastructure.Utilities;
+
+public class ValidationTypeUtility
+{
+    public static void ThrowIfNullOrDefault<T>(T value, string settingsName)
```
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Utilities/ValidationTypeUtility.cs:7]: string settingsName parameter shouldnt be for spesific. This is a utility function so we don't know what is the purpose of use of this method by programmers in the future. We cannot throw static message here

Change string settingsName to messageName and only throw the given messageName parameter as exception

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordAnswerRepository.cs (hunk 14 above)
Comment (Jerolivine) [inline Back End/MemorizeWords/MemorizeWords/Infrastructure/Persistance/Repository/WordAnswerRepository.cs:100]: WellDone for typecasting!

**Review by Jerolivine — CHANGES_REQUESTED** [review]

File: Back End/MemorizeWords-UnitTest/UnitTest.cs (hunk 11 above)
Comment (Jerolivine) [reply Back End/MemorizeWords-UnitTest/UnitTest.cs:92]: Can you add one more test for class type?

Comment (hamdiboyraz) [reply Back End/MemorizeWords-UnitTest/UnitTest.cs:92]: Done.

File: Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs (hunk 12 above)
Comment (hamdiboyraz) [reply Back End/MemorizeWords/MemorizeWords/Infrastructure/Extensions/IConfigurationExtension.cs:14]: Done.

**Pull request closed without merging**